examples/growing_bacteria.jl

```diff
@@ -0,0 +1,185 @@
+    end
+    
+    Agents.DBInterface.execute(model.space.updateq, (agent.pos..., agent.id))
+    return agent.pos
+end
+
+
+## Helper functions ########################################################################
+
+"""
+    interact!(a1::SimpleCell, a2::SimpleCell, model)
+Computes the interactions between two [`SimpleCell`](@ref)s using 
+[`noderepulsion`](@ref)
+"""
+function interact!(a1::SimpleCell, a2::SimpleCell, model)
+    
+    n11 = noderepulsion(a1.p1, a2.p1, model)
+    n12 = noderepulsion(a1.p1, a2.p2, model)
+    n21 = noderepulsion(a1.p2, a2.p1, model)
+    n22 = noderepulsion(a1.p2, a2.p2, model)
+   
+    a1.f1 = @. a1.f1 + (n11 + n12)
+    a1.f2 = @. a1.f2 + (n21 + n22)
+
+    a2.f1 = @. a2.f1 - (n11 + n21)
+    a2.f2 = @. a2.f2 - (n12 + n22)
+
+end
+
+
+"""
+    noderepulsion(p1, p2, model) → force
+Returns the repulsive force due to the interaction of nodes at `p1` and `p2`
+"""
+function noderepulsion(p1::NTuple{2, Float64}, p2::NTuple{2, Float64}, model::ABM)
+    delta = p1 .- p2
+    distance = norm(delta)
+
+    if distance ≤ 1
+        uv = delta./distance
+        return (model.properties[:hardness]*(1 - distance)).*uv
+    end
+
+    return (0, 0)
+end
+
+
+"""
+    transform_forces(agent::SimpleCell) → fsym, compression, torque
+Transforms the node forces into forces on the cell coordinates. 
+"""
+function transform_forces(agent::SimpleCell)
+    # symmetric forces (CM movement)
+    fsym = agent.f1 .+ agent.f2
+    # antisymmetric forces (compression, torque)
+    fasym = agent.f1 .- agent.f2
+    uv = unitvector(agent.orientation)
+    compression = dot(uv, fasym)
+    torque = 0.5*cross2D(uv, fasym)
+
+    return fsym, compression, torque
+end
+
+
+## simulation script #######################################################################
+
+space = ContinuousSpace(2, extend = (10, 10), periodic = false, metric = :euclidean)
+model = ABM(SimpleCell, space, properties = Dict(:dt => 0.005, :hardness => 1e2,
+                                                 :mobility => 1.0))
+
+add_agent!((5.0, 5.0), model, 0.0, 0.3, 0.0, 0.1)
+add_agent!((6.0, 5.0), model, 0.0, 0.0, 0.0, 0.1)
+
+
+nmax = 8000
+adata, = run!(model, agent_step!, model_step!, nmax;
+              adata = [:pos, :length, :orientation, :growthprog, :p1, :p2, :f1, :f2])
+
+
+#= 
+# temporary plotting function, uncomment for pretty pictures
+
+using PyPlot
+uniquecolor(id) = Tuple(matplotlib.colors.hsv_to_rgb(((id*Base.MathConstants.φ)%1, 0.8, 0.95)))
+for i ∈ 1:100:nmax
+    cla()
+    xlim(0, 10)
+    ylim(0, 10)
+    gca().set_aspect("equal")
+    data = sort!(adata[adata.step .== i, :], :id)
+    for r ∈ eachrow(data)
+        c1 = matplotlib.patches.Circle(r.p1, 0.5, color = uniquecolor(r.id), zorder=100)
+        c2 = matplotlib.patches.Circle(r.p2, 0.5, color = uniquecolor(r.id), zorder=100)
+   
+        gca().add_patch(c1)
+        gca().add_patch(c2)
+    end
+    title("n = "*string(i))
+    sleep(0.01)
+end
+=#
```

Comment:
Suggested plot replacement (`using Plots`)

```julia
result, = run!(model, agent_step!, model_step!, nmax; adata = [:p1, :p2])            
                      
theme(:vibrant)       
anim = @animate for i in 0:100:nmax  
    data = filter(row->row.step == i, result)           
    colors = HSV.(data[!, :id]*2.718.%1, data[!, :id]*3.14.%1, data[!, :id]*1.618.%1) 
    p1 = scatter(data[!, :p1], color=colors, lims=(0,10), markersize=15, legend=:none)
    scatter!(data[!, :p2], color=colors, markersize=15)                       
    title!(p1, "n = $(i)")                                  
end                                                                 
                                                                                                                                                                           
gif(anim, "bacteria.gif", fps = 25)
```

Reply:
![bacteria](https://user-images.githubusercontent.com/3059508/80286493-ff9c5380-872b-11ea-94e3-9b93bc1abdd0.gif)

Reply:
Isn't it possible to add lines connecting the balls? Even if they have the same color it doesn't feel like one bacterium,...

Reply:
Ah, I didn't even read the description properly. I can probably do one better than that...

Reply:
Don't have too much time today, but here's a proof of concept. Needs some resizing perhaps, and a rotation matrix applied to the markers still. 

Using `plotabm` :)

![bacteria](https://user-images.githubusercontent.com/3059508/80318420-495b6b80-880a-11ea-9d8e-8f3c72d0ad0e.gif)

Reply:
Holy crap this is amazing!

Reply:
```julia
## Build model #######################################################################

space = ContinuousSpace(2, extend = (10, 10), periodic = false, metric = :euclidean)
model = ABM(SimpleCell, space, properties = Dict(:dt => 0.005, :hardness => 1e2, 
                                                 :mobility => 1.0))

add_agent!((5.0, 5.0), model, 0.0, 0.3, 0.0, 0.1)
add_agent!((6.0, 5.0), model, 0.0, 0.0, 0.0, 0.1)

## Plotting ##########################################################################

function cassiniOval(agent)
    t = LinRange(0, 2π, 50)
    a = agent.growthprog
    b = 1
    m = @. 2*sqrt((b^4-a^4)+a^4*cos(2*t)^2)+2*a^2*cos(2*t)
    C = sqrt.(m/2)

    x = C.*cos.(t)
    y = C.*sin.(t)

    uv = unitvector(agent.orientation)
    θ = atan(uv[2],uv[1])
    R = [cos(θ) -sin(θ); sin(θ) cos(θ)]

    bacteria = R*permutedims([x y])
    Shape(bacteria[1,:], bacteria[2,:])
end

bacteria_colors(agent) = HSV.(agent.id*2.718.%1, agent.id*3.14.%1, agent.id*1.618.%1)

anim = @animate for i in 0:100:8000
    step!(model, agent_step!, model_step!, 100)
    p1 = plotabm(model, am = cassiniOval, as = 30, ac = bacteria_colors)
    title!(p1, "n = $(i)")
end

gif(anim, "bacteria.gif", fps = 25)
```

![bacteria](https://user-images.githubusercontent.com/3059508/80345811-04ffb800-886a-11ea-8589-2e480f1be5ad.gif)

Reply:
```julia
## Build model #######################################################################

space = ContinuousSpace(2, extend = (10, 10), periodic = false, metric = :euclidean)
model = ABM(SimpleCell, space, properties = Dict(:dt => 0.005, :hardness => 1e2, 
                                                 :mobility => 1.0))

add_agent!((5.0, 5.0), model, 0.0, 0.3, 0.0, 0.1)
add_agent!((6.0, 5.0), model, 0.0, 0.0, 0.0, 0.1)

## Plotting ##########################################################################

function cassiniOval(agent)
    t = LinRange(0, 2π, 50)
    a = agent.growthprog
    b = 1
    m = @. 2*sqrt((b^4-a^4)+a^4*cos(2*t)^2)+2*a^2*cos(2*t)
    C = sqrt.(m/2)

    x = C.*cos.(t)
    y = C.*sin.(t)

    uv = unitvector(agent.orientation)
    θ = atan(uv[2],uv[1])
    R = [cos(θ) -sin(θ); sin(θ) cos(θ)]

    bacteria = R*permutedims([x y])
    Shape(bacteria[1,:], bacteria[2,:])
end

bacteria_colors(agent) = HSV.(agent.id*2.718.%1, agent.id*3.14.%1, agent.id*1.618.%1)

anim = @animate for i in 0:100:8000
    step!(model, agent_step!, model_step!, 100)
    p1 = plotabm(model, am = cassiniOval, as = 30, ac = bacteria_colors)
    title!(p1, "n = $(i)")
end

gif(anim, "bacteria.gif", fps = 25)
```

![bacteria](https://user-images.githubusercontent.com/3059508/80345870-25c80d80-886a-11ea-9736-ac9dab60dc8c.gif)

Reply:
Awesome! Put this in the PR code!

Reply:
I'd also run it for more time. Its worth it!

Reply:
I may expand the space a bit too, since those guys like to run off outside of bounds.